torchtext/prototype/generate.py

```diff
@@ -0,0 +1,110 @@
+        """Greedy search decoding for text generation. Takes the most likely next token every time.
+
+        Inputs:
+            input_ids (Tensor): Text prompt(s) for greedy generation.
+            max_len (int): Max length to generate responses.
+            eos_idx (int): End of sequence index.
+            pad_idx (int): Padding index.
+            **model_kwargs
+        
+        Returns:
+            Batch of sequences decoded by greedy search.
+        """
+        unfinished_sequences = torch.ones((input_ids.shape[0], 1), device=input_ids.device, dtype=torch.long)
+
+        while True:
+            model_inputs = self.model.prepare_inputs_for_generation(input_ids, **model_kwargs)
+
+            # Get model output
+            outputs = self.model(**model_inputs)
+            decoder_output = outputs["decoder_output"]
+
+            # Calculate probabilities and take the most likely next token
+            probs = F.log_softmax(decoder_output[:, -1], dim=-1)
+            _, next_tokens = torch.topk(probs, 1)
+
+            # For any finished sequences, padding idx should be the last token
+            if eos_idx is not None:
+                if pad_idx is not None:
+                    next_tokens = next_tokens * unfinished_sequences + pad_idx * (1 - unfinished_sequences)
+
+            # Append the next tokens to the previous tokens
+            input_ids = torch.cat([input_ids, next_tokens], dim=-1)
+
+            if eos_idx is not None:
+                unfinished_sequences = unfinished_sequences.mul((next_tokens != eos_idx).long())
+
+            # Stop iterating once all sequences are finished or exceed the max_len
+            if unfinished_sequences.max() == 0 or len(input_ids[0]) >= max_len:
+                break
+
+        return model_inputs["decoder_tokens"]
+
+    def beam_search(self, input_ids: torch.Tensor, num_beams: int, max_len: Optional[int]) -> torch.Tensor:
+        raise NotImplementedError()
+
+    def generate(
+        self,
+        inputs: Optional[torch.Tensor] = None,
+        num_beams: Optional[int] = None,
+        max_len: Optional[int] = None,
+        eos_idx: int = 1,
+    ) -> torch.Tensor:
+        """Generating text via decoding for an encoder/decoder model.
+
+        Args:
+            input_ids (Tensor): Ids of transformed input tokens. The 'seed' text for generation.
+            num_beams (int): If provided, specifies the number of beams to use in beam search generation.
+            max_len (int): Max length to generate responses.
+            eos_idx (int): End of sequence index. Defaults to 1.
+
+        Returns:
+            Tensor of Tensors containing output sequences as ids.
+
+        `Note`: If one beam is provided or no beams are specified, the generation method will default to greedy search.
+        """
+        model_kwargs = {}
+
+        if self.is_encoder_decoder:
+            encoder = self.model.get_encoder()
+            model_kwargs["encoder_outputs"] = encoder(inputs)
```

Comment:
This should be the necessary args for the `forward` method of whatever model is being used in decoding.